### ISSUE TITLE
Request for additional test documentation

### ISSUE THREAD

**Issue by jkozol** [cockpit-project/cockpit#20138]
The testing framework is well documented on how tests run and how to test them. But, there is little/no documentation on the behaviour of the actual test commands.

A list of all current commands and how they work as well as recommendations on best practices for which commands should be used to reliably test.

**Comment by jelly** [comment]
Hi! Do you mean you want to have some online documentation for all our test commands like `machine.execute()` or `browser.click()`? All developer exposed functions in `testlib.py` should have a useful docstring, if not we should add those.

Are there in general methods you find confusing or which require more clarification?

**Comment by lucasgarfield** [comment]
@jelly, we're currently trying to upgrade to Patternfly 5 and we seem to be blocked by failing tests. @jkozol, maybe it would help the Cockpit team if you could provide some specific examples of tests that started failing after the PF5 dependency bump?

**Comment by jkozol** [comment]
I would appreciate any examples on the best ways to reliably select html elements in the testing environment.

**Comment by martinpitt** [comment]
@jkozol : Flakes *can* be related to bad CSS selectors -- exactly if a test fails with "selector is ambiguous". Then you need a more precise one. In most cases this should be a stable failure though which you can reproduce locally with `-stv` and `TEST_SHOW_BROWSER=1`, and then check in the inspector and devise a better selector. If it's a flake which is hard to reproduce, look at the test attachments, in particular the screenshot and the captured HTML -- there you should see the cause for the duplicate selector, like unexpectedly/temporarily rendering an item twice.

However, in my experience the vast majority of flakes are not related to bad selector, but simply that the expected and actual state are different.

There is no written documentation about "how to create a CSS selector", as that's a little bit of an art form. They should make sense to you as a human, be robust, and not too long. I. e. avoid things like "select the third column of the fourth row of the second table", as that is rather unstable. Instead, select by classes and data-specific identifiers. If necessary, add properties or classes to your React code. E.g. `table tr[data-machine-name='foo'] td.memusage`.

You can of course check our existing tests in all projects to get some inspiration how to select e.g. PatternFly elements. (`git grep .pf- test`). For that we won't create documentation, sorry -- all of these are too specific to pages, change a lot with PatternFly versions and redesigns in our pages, so they'd quickly go out of date and also not be very useful.

**Comment by lucasgarfield** [comment]
@martinpitt sorry for reviving this but I'm a bit puzzled, when I perform this test manually in my browser on Fedora and check the DOM, it is a `<li>`... what do you think might be causing it to render as a `<tr>` in RHEL? I don't know enough about how the Cockpit tests work to speculate - are they performed in a real browser?

**Comment by martinpitt** [comment]
@lucasgarfield Yes, they run in *the* real Fedora firefox and chromium packages. Did you run the actual integration test locally (with `TEST_SHOW_BROWSER=1 TEST_BROWSER=firefox` on the bots VM), or on a custom setup with manual steps?

Also, I don't know if the test actually fails in RHEL. So far I've only seen it fail on Fedora 38 and 39.

**Comment by martinpitt** [comment]
This was pretty much a support request, nothing further actionable here.
